File: sdk/python/examples/webhook_history.py
"""
Webhook history example — fetch recent webhook deliveries and inspect payload variants.

Usage:
    PACHCA_TOKEN=... python examples/webhook_history.py
"""

import asyncio
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "generated"))

from pachca.client import PachcaClient
from pachca.models import (
    ButtonWebhookPayload,
    ChatMemberWebhookPayload,
    CompanyMemberWebhookPayload,
    GetWebhookEventsParams,
    LinkSharedWebhookPayload,
    MessageWebhookPayload,
    ReactionWebhookPayload,
    ViewSubmitWebhookPayload,
    WebhookPayloadUnion,
)


def summarize_payload(payload: WebhookPayloadUnion) -> str:
    match payload:
        case LinkSharedWebhookPayload():
            return f"link_shared message_id={payload.message_id} links={len(payload.links)} user_id={payload.user_id}"
        case MessageWebhookPayload():
            return f"message event={payload.event} id={payload.id} chat_id={payload.chat_id}"
        case ReactionWebhookPayload():
            return f"reaction event={payload.event} message_id={payload.message_id} code={payload.code}"
        case ButtonWebhookPayload():
            return f"button message_id={payload.message_id} user_id={payload.user_id}"
        case ViewSubmitWebhookPayload():
            return f"view user_id={payload.user_id} fields={len(payload.data)}"
        case ChatMemberWebhookPayload():
            return f"chat_member event={payload.event} chat_id={payload.chat_id} users={len(payload.user_ids)}"
        case CompanyMemberWebhookPayload():
            return f"company_member event={payload.event} users={len(payload.user_ids)}"
        case _:
            return f"unknown type={type(payload).__name__}"


async def main():
    token = os.environ["PACHCA_TOKEN"]

    client = PachcaClient(token)
    response = await client.bots.get_webhook_events(GetWebhookEventsParams(limit=5))

    print(f"Fetched {len(response.data)} webhook events")
    for index, event in enumerate(response.data, start=1):
        print(
            f"{index}. id={event.id} created_at={event.created_at.isoformat()} payload={summarize_payload(event.payload)}"
        )

    print(
        f'has_next={response.meta.paginate.has_next} next_page="{response.meta.paginate.next_page}"'
    )

    await client.close()


asyncio.run(main())

File: sdk/typescript/examples/webhook-history.ts
/**
 * Webhook history example — fetch recent webhook deliveries and inspect payload variants.
 *
 * Usage:
 *   PACHCA_TOKEN=your_token bun run examples/webhook-history.ts
 */

import { PachcaClient } from "../src/index.js";
import type {
  ButtonWebhookPayload,
  ChatMemberWebhookPayload,
  CompanyMemberWebhookPayload,
  LinkSharedWebhookPayload,
  MessageWebhookPayload,
  ReactionWebhookPayload,
  ViewSubmitWebhookPayload,
  WebhookPayloadUnion,
} from "../src/generated/types.js";

const token = process.env.PACHCA_TOKEN;
if (!token) {
  console.error("Set PACHCA_TOKEN environment variable");
  process.exit(1);
}

const client = new PachcaClient(token);
const response = await client.bots.getWebhookEvents({ limit: 5 });

console.log(`Fetched ${response.data.length} webhook events`);
for (const [index, event] of response.data.entries()) {
  console.log(
    `${index + 1}. id=${event.id} created_at=${event.createdAt} payload=${summarizePayload(event.payload)}`,
  );
}

console.log(
  `has_next=${response.meta.paginate.hasNext} next_page=${JSON.stringify(response.meta.paginate.nextPage)}`,
);

function summarizePayload(payload: WebhookPayloadUnion): string {
  if (payload.type === "message" && payload.event === "link_shared") {
    const linkShared = payload as LinkSharedWebhookPayload;
    return `link_shared message_id=${linkShared.messageId} links=${linkShared.links.length} user_id=${linkShared.userId}`;
  }
  switch (payload.type) {
    case "message": {
      const message = payload as MessageWebhookPayload;
      return `message event=${message.event} id=${message.id} chat_id=${message.chatId}`;
    }
    case "reaction": {
      const reaction = payload as ReactionWebhookPayload;
      return `reaction event=${reaction.event} message_id=${reaction.messageId} code=${reaction.code}`;
    }
    case "button": {
      const button = payload as ButtonWebhookPayload;
      return `button message_id=${button.messageId} user_id=${button.userId}`;
    }
    case "view": {
      const view = payload as ViewSubmitWebhookPayload;
      return `view user_id=${view.userId} fields=${Object.keys(view.data).length}`;
    }
    case "chat_member": {
      const member = payload as ChatMemberWebhookPayload;
      return `chat_member event=${member.event} chat_id=${member.chatId} users=${member.userIds.length}`;
    }
    case "company_member": {
      const member = payload as CompanyMemberWebhookPayload;
      return `company_member event=${member.event} users=${member.userIds.length}`;
    }
    default:
      return "unknown";
  }
}
